File: app/src/main/java/com/cyl/musiclake/ui/music/playlist/PlaylistDetailPresenter.java
package com.cyl.musiclake.ui.music.playlist;

import com.cyl.musiclake.RxBus;
import com.cyl.musiclake.api.MusicApiServiceImpl;
import com.cyl.musiclake.api.PlaylistApiServiceImpl;
import com.cyl.musiclake.base.BasePresenter;
import com.cyl.musiclake.bean.Music;
import com.cyl.musiclake.bean.Playlist;
import com.cyl.musiclake.event.PlaylistEvent;
import com.cyl.musiclake.net.ApiManager;
import com.cyl.musiclake.net.RequestCallBack;
import com.cyl.musiclake.utils.LogUtil;
import com.cyl.musiclake.utils.ToastUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;


/**
 * Created by yonglong on 2018/1/7.
 */

public class PlaylistDetailPresenter extends BasePresenter<PlaylistDetailContract.View> implements PlaylistDetailContract.Presenter {
    private static final String TAG = "PlaylistDetailPresenter";
    private List<String> netease = new ArrayList<>();
    private List<String> qq = new ArrayList<>();
    private List<String> xiami = new ArrayList<>();

    @Inject
    public PlaylistDetailPresenter() {
    }

    @Override
    public void loadPlaylistSongs(String playlistID) {
        ApiManager.request(PlaylistApiServiceImpl.INSTANCE.getMusicList(playlistID), new RequestCallBack<List<Music>>() {
            @Override
            public void success(List<Music> result) {
//                mView.showPlaylistSongs(result);
                netease.clear();
                qq.clear();
                xiami.clear();
                for (Music music : result) {
                    if (music.getType() == Music.Type.NETEASE) {
                        netease.add(music.getId());
                    } else if (music.getType() == Music.Type.QQ) {
                        qq.add(music.getId());
                    } else if (music.getType() == Music.Type.XIAMI) {
                        xiami.add(music.getId());
                    }
                }
                getBatchSongDetail("netease", netease.toArray(new String[netease.size()]));
                getBatchSongDetail("qq", qq.toArray(new String[qq.size()]));
                getBatchSongDetail("xiami", xiami.toArray(new String[xiami.size()]));
            }

            @Override
            public void error(String msg) {
                LogUtil.e(TAG, msg);
                ToastUtils.show(msg);
            }
        });
    }

    @Override
    public void deletePlaylist(Playlist playlist) {
        ApiManager.request(PlaylistApiServiceImpl.INSTANCE.deletePlaylist(playlist.getId()), new RequestCallBack<String>() {
            @Override
            public void success(String result) {
                mView.success(1);
                RxBus.getInstance().post(new PlaylistEvent());
                ToastUtils.show(result);
            }

            @Override
            public void error(String msg) {
                ToastUtils.show(msg);
            }
        });
    }

    @Override
    public void renamePlaylist(Playlist playlist, String title) {
        ApiManager.request(PlaylistApiServiceImpl.INSTANCE.renamePlaylist(playlist.getId(), title), new RequestCallBack<String>() {
            @Override
            public void success(String result) {
                mView.success(1);
                RxBus.getInstance().post(new PlaylistEvent());
                ToastUtils.show(result);
            }

            @Override
            public void error(String msg) {
                ToastUtils.show(msg);
            }
        });
    }

    @Override
    public void disCollectMusic(String pid, int position, Music music) {
        ApiManager.request(PlaylistApiServiceImpl.INSTANCE.disCollectMusic(pid, music), new RequestCallBack<String>() {
            @Override
            public void success(String result) {
                mView.removeMusic(position);
            }

            @Override
            public void error(String msg) {
                ToastUtils.show(msg);
            }
        });
    }

    public void getBatchSongDetail(String vendor, String[] ids) {
        ApiManager.request(MusicApiServiceImpl.INSTANCE.getBatchMusic(vendor, ids), new RequestCallBack<List<Music>>() {
            @Override
            public void success(List<Music> result) {
                mView.showPlaylistSongs(result);
            }

            @Override
            public void error(String msg) {
                ToastUtils.show(msg);
            }
        });
    }
}

File: app/src/main/java/com/cyl/musiclake/ui/music/playlist/PlaylistDetailPresenter.kt
package com.cyl.musiclake.ui.music.playlist

import com.cyl.musiclake.api.MusicApiServiceImpl
import com.cyl.musiclake.api.PlaylistApiServiceImpl
import com.cyl.musiclake.api.baidu.BaiduApiServiceImpl
import com.cyl.musiclake.api.netease.NeteaseApiServiceImpl
import com.cyl.musiclake.base.BasePresenter
import com.cyl.musiclake.bean.Album
import com.cyl.musiclake.bean.Artist
import com.cyl.musiclake.bean.Music
import com.cyl.musiclake.bean.Playlist
import com.cyl.musiclake.common.Constants
import com.cyl.musiclake.data.PlaylistLoader
import com.cyl.musiclake.data.SongLoader
import com.cyl.musiclake.event.MyPlaylistEvent
import com.cyl.musiclake.net.ApiManager
import com.cyl.musiclake.net.RequestCallBack
import com.cyl.musiclake.utils.LogUtil
import com.cyl.musiclake.utils.ToastUtils
import org.greenrobot.eventbus.EventBus
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import java.util.*
import javax.inject.Inject


/**
 * Created by yonglong on 2018/1/7.
 */

class PlaylistDetailPresenter @Inject
constructor() : BasePresenter<PlaylistDetailContract.View>(), PlaylistDetailContract.Presenter {

    private val netease = ArrayList<String>()
    private val qq = ArrayList<String>()
    private val xiami = ArrayList<String>()

    override fun loadArtistSongs(artist: Artist) {
        if (artist.type == null || artist.type == Constants.LOCAL) {
            doAsync {
                val data = SongLoader.getSongsForArtist(artist.name)
                uiThread {
                    mView.showPlaylistSongs(data)
                }
            }
            return
        } else if (artist.type == Constants.BAIDU) {
            val observable = BaiduApiServiceImpl.getArtistSongList(artist.artistId.toString(), 0)
            ApiManager.request(observable, object : RequestCallBack<Artist> {
                override fun success(result: Artist) {
                    mView?.showPlaylistSongs(result.songs)
                }

                override fun error(msg: String) {
                    LogUtil.e(TAG, msg)
                    mView?.showError(msg, true)
                    ToastUtils.show(msg)
                }
            })
            return
        }
        val observable = MusicApiServiceImpl.getArtistSongs(artist.type!!, artist.artistId.toString(), 50, 0)
        ApiManager.request(observable, object : RequestCallBack<Artist> {
            override fun success(result: Artist) {
                val musicLists = result.songs
                val iterator = musicLists.iterator()
                while (iterator.hasNext()) {
                    val temp = iterator.next()
                    if (temp.isCp) {
                        //list.remove(temp);// 出现java.util.ConcurrentModificationException
                        iterator.remove()// 推荐使用
                    }
                }
                mView?.showPlaylistSongs(musicLists)
            }

            override fun error(msg: String) {
                LogUtil.e(TAG, msg)
                mView?.showError(msg, true)
                ToastUtils.show(msg)
            }
        })
    }

    override fun loadAlbumSongs(album: Album) {
        if (album.type == null || album.type == Constants.LOCAL) {
            doAsync {
                val data = SongLoader.getSongsForAlbum(album.name)
                uiThread {
                    mView.showPlaylistSongs(data)
                }
            }
            return
        } else if (album.albumId == null) {
            mView?.showPlaylistSongs(null)
            return
        } else if (album.type == Constants.BAIDU) {
            val observable = BaiduApiServiceImpl.getAlbumSongList(album.albumId.toString())
            ApiManager.request(observable, object : RequestCallBack<Album> {
                override fun success(result: Album) {
                    mView?.showPlaylistSongs(result.songs)
                }

                override fun error(msg: String) {
                    LogUtil.e(TAG, msg)
                    mView?.showError(msg, true)
                    ToastUtils.show(msg)
                }
            })
            return
        }
        val observable = MusicApiServiceImpl.getAlbumSongs(album.type.toString(), album.albumId.toString())
        ApiManager.request(observable, object : RequestCallBack<MutableList<Music>> {
            override fun success(result: MutableList<Music>) {
                val iterator = result.iterator()
                while (iterator.hasNext()) {
                    val temp = iterator.next()
                    if (temp.isCp) {
                        //list.remove(temp);// 出现java.util.ConcurrentModificationException
                        iterator.remove()// 推荐使用
                    }
                }
                mView?.showPlaylistSongs(result)
            }

            override fun error(msg: String) {
                LogUtil.e(TAG, msg)
                mView?.showError(msg, true)
                ToastUtils.show(msg)
            }
        })
    }


    override fun loadPlaylistSongs(playlist: Playlist) {
        when (playlist.type) {
            Constants.PLAYLIST_LOCAL_ID,
            Constants.PLAYLIST_HISTORY_ID,
            Constants.PLAYLIST_LOVE_ID,
            Constants.PLAYLIST_QUEUE_ID -> doAsync {
                val data = playlist.pid?.let { PlaylistLoader.getMusicForPlaylist(it, playlist.order) }
                uiThread {
                    if (data != null && data.isNotEmpty()) {
                        mView?.showPlaylistSongs(data)
                    } else {
                        mView?.showEmptyState()
                    }
                }
            }
            Constants.PLAYLIST_BD_ID -> {
                ApiManager.request(BaiduApiServiceImpl.getRadioChannelInfo(playlist), object : RequestCallBack<Playlist> {
                    override fun error(msg: String?) {
                        mView?.showError(msg, true)
                    }

                    override fun success(result: Playlist?) {
                        result?.let {
                            if (it.musicList.isNotEmpty()) {
                                mView?.showPlaylistSongs(it.musicList)
                            } else {
                                mView?.showEmptyState()
                            }
                        }
                    }

                })

            }
            Constants.PLAYLIST_WY_ID -> {
                ApiManager.request(playlist.pid?.let { NeteaseApiServiceImpl.getPlaylistDetail(it) }, object : RequestCallBack<Playlist> {
                    override fun error(msg: String?) {
                        mView?.showError(msg, true)
                    }

                    override fun success(result: Playlist?) {
                        result?.let {
                            if (it.musicList.isNotEmpty()) {
                                mView?.showPlaylistSongs(it.musicList)
                            } else {
                                mView?.showEmptyState()
                            }
                        }
                    }

                })

            }
            else -> ApiManager.request(playlist.pid?.let { PlaylistApiServiceImpl.getMusicList(it) }, object : RequestCallBack<MutableList<Music>> {
                override fun success(result: MutableList<Music>) {
                    val iterator = result.iterator()
                    while (iterator.hasNext()) {
                        val temp = iterator.next()
                        if (temp.isCp) {
                            iterator.remove()// 推荐使用
                        }
                    }
                    mView?.showPlaylistSongs(result)
                }

                override fun error(msg: String) {
                    LogUtil.e(TAG, msg)
                    mView?.showError(msg, true)
                    ToastUtils.show(msg)
                }
            })
        }
    }

    /**
     * 删除歌单
     */
    override fun deletePlaylist(playlist: Playlist) {

    }

    override fun renamePlaylist(playlist: Playlist, title: String) {
        ApiManager.request(playlist.pid?.let { PlaylistApiServiceImpl.renamePlaylist(it, title) }, object : RequestCallBack<String> {
            override fun success(result: String) {
                mView.success(1)
                playlist.name = title
                EventBus.getDefault().post(MyPlaylistEvent(Constants.PLAYLIST_RENAME, playlist))
                ToastUtils.show(result)
            }

            override fun error(msg: String) {
                ToastUtils.show(msg)
            }
        })
    }

    companion object {
        private val TAG = "PlaylistDetailPresenter"
    }
}
